augur/io/metadata.py

```diff
@@ -127,4 +142,32 @@ def read_metadata_to_dict(metadata):
                 "Metadata file must be a CSV or TSV file."
             ) from err
 
-        yield from csv.DictReader(handle, dialect=dialect)
+        metadata_reader = csv.DictReader(handle, dialect=dialect)
+        if duplicate_reporting is DataErrorMethod.SILENT:
+            # Directly yield from metadata reader since we do not need to check for duplicate ids
+            yield from metadata_reader
+        else:
+            if id_column is None:
+                id_column = metadata_reader.fieldnames[0]
+
+            for record in metadata_reader:
+                record_id = record.get(id_column)
+                if record_id is None:
+                    raise AugurError(f"The provided id column {id_column!r} does not exist in the metadata.")
+
+                if record_id in seen_ids:
+                    duplicate_ids.add(record_id)
+                else:
+                    seen_ids.add(record_id)
+
+                yield record
```

Comment:
Hmm, so should we raise the one dup error and ask the user to make sure they don't have any other duplicates? 
I personally think it would be frustrating to correct one dup error then hit another one in the next run.

Reply:
That's a good point. Agreed it can be frustrating to be stuck in an error correction loop. OTOH, it can also be frustrating if something seems to be processing data fine and then only spits out errors at the end, after you've waited around for it.

(If the process is never slow, then that's fine as the iteration time is still short. But I suspect we won't always be fast here.)

Reply:
Yeah I can see both sides. I can split the `ERROR` option into something like `ERROR_FIRST` and `ERROR_ALL` to support both ways.

Reply:
Sounds good to me. Either would be a good default. (I'd probably choose `ERROR_FIRST`.)

Reply:
Things seem a little complicated because the two error methods, but I think still manageable.
See the different warnings/errors in d56388295fcc804fbe42d474b2e3636f92b251e5, 8e1856ff28be69bc2c42f3eff2e4e4c68e2553ea, and 0b637d298c9b90f3f76f82f00af4e806bb63b960

Reply:
:raised_hands: Seems manageable to me. Good error handling is always a bit complicated to implement, but IME it pays off a lot for end users (which means ourselves too :-).